File: src/algorithms/tokenBucket.ts
import { Bucket } from "../types/bucket";
import { ClientConfig } from "../types/client";
import { ConsumeResult } from "../types/result";

export function consume(
    bucket: Bucket,
    clientConfig: ClientConfig
): ConsumeResult {

    const currentTime = Date.now();

    const refilledBucket = refill(bucket, clientConfig, currentTime);

    const allowed = canConsume(refilledBucket);

    const updatedBucket = allowed
        ? consumeToken(refilledBucket)
        : refilledBucket;

    return {
        allowed,
        bucket: updatedBucket,
        remainingTokens: updatedBucket.tokens,
        resetTime: calculateResetTime(
            updatedBucket,
            clientConfig,
            currentTime
        )
    };
}

function refill(
    bucket: Bucket,
    clientConfig: ClientConfig,
    currentTime: number
): Bucket {

    const elapsedTime = currentTime - bucket.lastRefillTime;
    const elapsedSeconds = elapsedTime / 1000;

    const tokensToAdd = Math.floor(
        elapsedSeconds * clientConfig.refillRate
    );

    if (tokensToAdd <= 0) {
        return bucket;
    }

    return {
        ...bucket,
        tokens: Math.min(
            bucket.tokens + tokensToAdd,
            clientConfig.capacity
        ),
        lastRefillTime: currentTime
    };
}

function canConsume(bucket: Bucket): boolean {
    return bucket.tokens > 0;
}

function consumeToken(bucket: Bucket): Bucket {
    return {
        ...bucket,
        tokens: bucket.tokens - 1
    };
}

function calculateResetTime(
    bucket: Bucket,
    clientConfig: ClientConfig,
    currentTime: number
): number {

    if (bucket.tokens > 0) {
        return currentTime;
    }

    const millisecondsPerToken = 1000 / clientConfig.refillRate;

    return bucket.lastRefillTime + millisecondsPerToken;
}

File: src/lua/tokenBucket.lua
local bucketKey = KEYS[1]

local capacity = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[2])
local currentTime = tonumber(ARGV[3])

----------------------------------------------------
-- Fetch bucket
----------------------------------------------------

local bucket = redis.call(
    "HMGET",
    bucketKey,
    "tokens",
    "lastRefillTime"
)

local tokens = tonumber(bucket[1])
local lastRefillTime = tonumber(bucket[2])

if not tokens or not lastRefillTime then
    tokens = capacity
    lastRefillTime = currentTime
end

----------------------------------------------------
-- Refill
----------------------------------------------------

local elapsedTime = currentTime - lastRefillTime
local elapsedSeconds = elapsedTime / 1000

local tokensToAdd = math.floor(
    elapsedSeconds * refillRate
)

if tokensToAdd > 0 then

    tokens = math.min(
        tokens + tokensToAdd,
        capacity
    )

    local millisecondsPerToken =
        1000 / refillRate

    lastRefillTime =
        lastRefillTime +
        (tokensToAdd * millisecondsPerToken)
end

----------------------------------------------------
-- Consume
----------------------------------------------------

local allowed = 0

if tokens > 0 then
    tokens = tokens - 1
    allowed = 1
end

----------------------------------------------------
-- Save Bucket
----------------------------------------------------

redis.call(
    "HSET",
    bucketKey,
    "tokens",
    tokens,
    "lastRefillTime",
    lastRefillTime
)

redis.call(
    "EXPIRE",
    bucketKey,
    3600
)

----------------------------------------------------
-- Reset Time
----------------------------------------------------

local resetTime

if tokens > 0 then

    resetTime = currentTime

else

    local millisecondsPerToken =
        1000 / refillRate

    resetTime =
        lastRefillTime +
        millisecondsPerToken

end

----------------------------------------------------

return {
    allowed,
    tokens,
    resetTime
}